### DISCUSSION TITLE
Acessibilidade com React/styled-components

### DISCUSSION THREAD

**Discussion by doglasbatista** [frontendbr/forum#1859] [Discussão]
Olá...

Estou no começo do desenvolvimento de uma aplicação utilizando React com styled-components e gostaria de recomendações sobre cuidados, ferramentas e qualquer coisa que possa me ajudar e tomar cuidado com as questões de acessibilidade dessa aplicação

**Comment by ninetails** [comment]
Opa @doglasbatista , primeiramente parabéns por já pensar em acessibilidade antes mesmo de botar a mão na massa.

Opinando sobre a _stack_, creio que o _styled-components_ não influirá em relação ao tema e sobre o _React_, não terá problemas pois a biblioteca por padrão repassa as _props_ como atributos pro _frontend_.

Para auxiliar no desenvolvimento, cito 2 utilizações do [axe](https://www.deque.com/axe/):
- [jest-axe](https://github.com/nickcolley/jest-axe) para validar acessibilidade durante os testes
- [react-axe](https://github.com/dequelabs/react-axe) para ver erros de acessibilidade no _console_

Também existem outras formas de testar, o _axe_ em si ele vai checar partes do teu código e não o código inteiro da página. E também não lembro se o _axe_ valida cores (questões como contrastes e tipos de daltonismo). Logo sempre é bom se preocupar em validar o HTML da página como por exemplo utilizando o [validador da w3c](https://validator.w3.org/) para checar semântica. Já que citou o _styled-components_, vale também ressaltar que existe um plugin pra validar _CSS_ [utilizando o stylelint](https://styled-components.com/docs/tooling#stylelint) caso também se preocupe com semântica no _CSS_.

Já sobre checagem que falei anteriormente como cores, contraste e daltonismo, se preocupando com WAI-ARIA, microformats... só _googlando_ mesmo para achar outras ferramentas e _bookmarklets_.

**Comment by doglasbatista** [comment]
@ninetails Cara... Valeuzão pelas dicas, vou dar uma olhada com calma nos link....

Sobre o styled-components... Vacilo meu de não ter dado mais contexto do pq dele na pergunta.

A gente ta usando eslint no projeto, e ele tem esse plugin para acessibilidade, o [eslint-plugin-jsx-a11y](https://www.npmjs.com/package/eslint-plugin-jsx-a11y). Porém ele não funciona muito bem com styled-components. Tanto que o pessoal acabou criando o [eslint-plugin-styled-components-a11y](https://www.npmjs.com/package/eslint-plugin-styled-components-a11y). Mas ele ainda está no começo

**Comment by ninetails** [comment]
Opa @doglasbatista não manjava que o _jsx-a11y_ tinha problemas com _styled-components_... Sou daqueles que considera "quanto mais tooling, melhor" rs. Porém creio que deixar a checagem após a renderização possa dar uma maior segurança, já que nisso o _axe_ ajuda bastante.

No que digo em relação a validação _pós-output_ é pq devido ao React refletir _props_ pro _frontend_, é que é sempre bom ver se não está indo _props_ pro HTML. Não em relação a acessibilidade mas relacionado a semântica, utilizando _styled-components_ pode ter que fazer coisas do tipo:

```jsx
const StyledDiv = styled(({propIndesejadaUsadaNoSC, ...props}) => <div {...props} />)`
  regra: ${({propIndesejadaUsadaNoSC}) => `atributo css que usa ${propIndesejadaUsadaNoSC}`};
`
```

Se usar direto o `styled.div`, poderia ter um output do tipo `<div propIndesejadaUsadaNoSC="algum valor">...</div>`

**update:** Ou melhor ainda, quando eu tinha visto sobre o assunto ainda não tinha o [shouldForwardProp](https://github.com/styled-components/styled-components/releases/tag/v5.1.0) pro SC, só pro emotion... muito menos [transient props :heart_eyes:](https://medium.com/@probablyup/introducing-transient-props-f35fd5203e0c).

**Comment by drianoaz** [comment]
Estava pesquisando sobre acessibilidade e encontrei uma ferramenta que achei sensacional. [O Storybook possui um Addon chamado a11y.](https://github.com/storybookjs/storybook/tree/master/addons/a11y) Ele basicamente roda os mesmos testes de acessibilidade do Google Lighthouse dentro do Storybook.

O Storybook é uma ferramenta que já faz parte do nosso workflow. **Em todos esses anos, eu simplesmente não sabia que isso existia. Saber dessa existência, mudou a forma vou trabalhar de hoje em diante.** 

É incrível como isso está na página home do storybook e mesmo assim todas as equipes que trabalhei sempre ignoraram essa ferramenta. (Vai ver precisamos fomentar o uso dela)

> With a11y you can test compliance of your stories with web accessibility standards.

> The “A11y Addon” for Storybook runs on top of the Axe Accessibility Engine developed by Deque Systems. This is the same engine used by Google Lighthouse to test and score entire web pages.

### Como isso funciona na prática?

Quando criamos um story, podemos importar esse addon e ao fazer isso, uma nova aba será inserida e nela será exibida algumas informações bem interessantes:

<img src="https://user-images.githubusercontent.com/15058771/88446784-5c7ad900-ce03-11ea-81cb-f13331442bb2.png" width="500px" />

Eu encontrei um [artigo bem bacana falando um pouco sobre esse addon](https://medium.com/storybookjs/instant-accessibility-qa-linting-in-storybook-4a474b0f5347)

### Beleza, mas isso roda no storybook e não é um teste automatizado

Sem muito esforço, é possível se criar uma cultura de code-review que verifique se novos componentes estão passando nesse teste, mesmo que manualmente. 

Pela natureza do storybook, é possível adicionar uma etapa no CI para compilar e publicar o storybook em uma URL temporária para que as pessoas que vão fazer o code-review não tenham que rodar na máquina delas.

### Qual é o ganho real disso?

Já pensou que louco ter um playground onde qualquer membro da sua equipe consegue consultar todos os componentes existentes na sua aplicação e no mesmo lugar saber se eles são acessíveis? Isso pode trazer melhorias valiosas para qualquer produto.
